pylops/optimization/solver.py

```diff
@@ -2,10 +2,202 @@
 
 import numpy as np
 
+from pylops.optimization.basesolver import Solver
 from pylops.utils.backend import get_array_module
 
 
-def cg(Op, y, x0, niter=10, damp=0.0, tol=1e-4, show=False, callback=None):
+class CG(Solver):
+    r"""Conjugate gradient
+
+    Solve a square system of equations given an operator ``Op`` and
+    data ``y`` using conjugate gradient iterations.
+
+    Parameters
+    ----------
+    Op : :obj:`pylops.LinearOperator`
+        Operator to invert of size :math:`[N \times N]`
+    y : :obj:`np.ndarray`
+        Data of size :math:`[N \times 1]`
+
+    Attributes
+    ----------
+    x : :obj:`np.ndarray`
+        Estimated model of size :math:`[N \times 1]`
+    iit : :obj:`int`
+        Number of executed iterations
+    cost : :obj:`numpy.ndarray`, optional
+        History of the L2 norm of the residual
+
+    Notes
+    -----
+    Solve the :math:`\mathbf{y} = \mathbf{Opx}` problem using conjugate gradient
+    iterations [1]_.
+
+    .. [1] Hestenes, M R., Stiefel, E., “Methods of Conjugate Gradients for Solving
+       Linear Systems”, Journal of Research of the National Bureau of Standards.
+       vol. 49. 1952.
+
+    """
+
+    def __init__(self, Op, y):
+        self.Op = Op
+        self.y = y
+        self.ncp = get_array_module(y)
+
+    def _print_setup(self):
+        self.tstart = time.time()
+        if self.niter is None:
+            strpar = "tol = {self.tol:10e}\tniter = {self.niter}"
+        else:
+            strpar = "tol = {self.tol:10e}"
+        print(
+            "CG\n"
+            "-----------------------------------------------------------\n"
+            f"The Operator Op has {self.Op.shape[0]} rows and {self.Op.shape[1]} cols\n"
+        )
+        print(strpar)
+        print("-----------------------------------------------------------")
+        head1 = "    Itn           x[0]              r2norm"
+        print(head1)
+
+    def _print_step(self):
+        if not np.iscomplex(self.x[0]):
+            msg = f"{self.iiter:6g}        {self.x[0]:11.4e}        {self.cost[self.iiter]:11.4e}"
+        else:
+            msg = f"{self.iiter:6g}    {np.real(self.x[0]):4.1e} + {np.imag(self.x[0]):4.1e}j    {self.cost[self.iiter]:11.4e}"
+        print(msg)
+
+    def _print_finalize(self):
+        print(
+            f"\nIterations = {self.iiter}        Total time (s) = {time.time() - self.tstart:.2f}"
+        )
+        print("-----------------------------------------------------------------\n")
+
+    def setup(self, x0=None, niter=None, tol=1e-4, show=False):

```

Comment:
Part of this can also be in the base class, and then just call super() for the implementation of CG

Reply:
_print_finalize is now also in the base call.

I am not sure I understand why we need to call super() though

Reply:
_print_finalize is now also in the base call.

I am not sure I understand why we need to call super() though

Reply:
Well, the way it is defined now that won't be necessary.

Reply:
ok, then you need to explain better as I don't understand... please provide a short example of what you mean

Reply:
Sorry, I was confused that you mentioned `_print_finalize`, but the original comment was regarding `setup()`.
My idea, as I recall it now, was to move the common parts for solvers, e.g. initialization of `x0`, `self.cost` and the starting value, `self.iiter` to the base class, then call `super()` ad add the solver specific parameters.

Reply:
Oh right, I see now. 

The idea of setup is to avoid the __init__ doing too much. For example for a solver like ISTA I foresee the computation of the max eigenvalue to be more suited for setup than for the init method (as a matter of preference I would like init to be almost instantaneous). So we could define some 'super'-hyperparameters that are always present (eg niter) and move them to the init, would this be better for you?

Reply:
I agree with this. The init should only register the class attributes. Any real computation should be elsewhere (i.e., setup).

Reply:
Oook, then I think I am going to resolve this one. Having now worked on refractoring 3 solvers, I think the only 3 things that all solvers have in common are operator, data, and initial model... as we want the latter two to be passed at setup and solve stage...

I think also for solvers like FISTA, setup will trigger the computation of the Lipschitz constant and if we call again setup/solve (with for example a new y) we can simply check that if it is already being computed and stored we just don't do it again.